### PR TITLE
Add role-based access checks to the frontend

### LINKED ISSUE #27: Add access checks to frontend app
This will need to be done after proper JWTs and bearer token authentication are implemented.

The frontend should be able to read the roles of the authenticated user (via the auth token) and then conditionally hide sections in the nav bar or block access to pages based on those roles.

The Admin section in the nav bar should only be visible to users with the 'Admin' role.
If a non-admin user attempts to route to the admin page, they will be returned to the login page, and a toast error will be displayed indicating that they do not have access.

### PR DESCRIPTION
**Pull request by ginderjeremiah**
## Summary

Implements **#27** — the frontend now reads the authenticated user's roles from the JWT and gates the admin UI on the `Admin` role.

The prerequisite (proper JWT/bearer auth) called out in the issue is satisfied: it shipped in **#14** (backend auth modernization) and **#47** (frontend token persistence), both closed. The access token already carries `role` claims; this PR consumes them.

## How it addresses the issue

- **Read roles from the auth token** (pure, framework-agnostic, in `$lib/api`):
  - `token-store.ts` — refactored the JWT-payload decode into one helper and added `getRoles()`, normalizing both the single-role (string) and multi-role (array) JWT shapes.
  - `roles.ts` (new) — `ERole` enum mirroring backend `Game.Core.ERole` (hand-maintained domain concept, not a codegen'd DTO) + `hasRole()` predicate.
- **Hide the Admin nav section for non-admins:**
  - `screens/screen-defs.ts` (new) — moved the screen list out of the game page, marked Admin with `requiresRole: ERole.Admin`, and added a pure `visibleScreens()` filter.
  - `game/+page.svelte` — renders `visibleScreens(GAME_SCREENS, roles)`. Roles are snapshotted on mount so the gating matches the server render and updates on the client with **no hydration mismatch** (mirrors the boot gate's `booting`/`hydrated` pattern). The existing `NavSidebar` already collapses a group with no items, so the whole Admin group disappears.
- **Block non-admins from `/admin`:**
  - `admin/admin-access.ts` (new) — `ensureAdminAccess()`: when the Admin role is absent, surfaces a `toastError` and returns the user to the login page via `goto('/')` (SPA nav, so the toast survives), returning `false`.
  - `admin/+page.svelte` — guards on mount and gates rendering behind an `authorized` flag, so a non-admin never mounts the workbench or triggers its reference load.

This is the **UX layer only** — the backend independently enforces the Admin role on every admin endpoint (`AdminRoleAuthorizationFilter`), which remains the real security boundary.

## Test plan

- [x] `npm run check` — 0 errors / 0 warnings
- [x] `npm run lint` — clean (incl. `e2e-tests`)
- [x] `npm run test:unit` (vitest) — **544 passing**, including new/updated:
  - `token-store.test.ts` — `getRoles` (single/array/none/non-JWT/logged-out, ignores non-string array entries)
  - `roles.test.ts` (new) — `hasRole`
  - `screen-defs.test.ts` (new) — `visibleScreens` gating + `GAME_SCREENS` shape
  - `admin-access.test.ts` (new) — guard grants admins, toasts + redirects non-admins
  - `NavSidebar.test.ts` — Admin group/item absent when the screen is filtered out
- [x] e2e seed trigger validated against Postgres (admin marker → role granted; normal account → none; idempotent re-run)

### E2E note
The admin area is now role-gated, but the e2e suite creates fresh (non-admin) accounts. Hand-seeding an admin user is brittle (custom password hash + full player graph). Instead, an **e2e-only** trigger in `e2e-seed.sql` grants the seeded `Admin` role to any account whose username starts with `e2eadmin`, and a new `createAdminAndStartGame()` helper uses that marker. Each admin test still creates a unique account, so they stay parallel-safe. Added a `navigation.test.ts` case asserting the Admin nav entry is hidden for a normal player.

## Docs

Added a "Role-based access checks" note to the Authentication section of `docs/frontend.md`.

## Notes / observations

- `ScreenDef` is intentionally duplicated between the shared `NavSidebar` component (its prop contract) and `screen-defs.ts` to avoid coupling a shared component to a route — left as-is.

Closes #27

---
_Generated by [Claude Code](https://claude.ai/code/session_01H4PetKqT2pTdFpnQWY12Kd)_

### CODE REVIEW

File: UI/new-svelte/src/routes/admin/admin-access.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import { goto } from '$app/navigation';
+import { resolve } from '$app/paths';
+import { ERole, hasRole } from '$lib/api';
+import { toastError } from '$stores';
+
+/**
+ * Client-side guard for the admin area. Grants access only to holders of the {@link ERole.Admin}
+ * role; otherwise it surfaces an access-denied toast and returns the user to the login page, then
+ * reports that access was denied so the caller can keep the privileged UI from rendering.
```
Comment (ginderjeremiah) [inline UI/new-svelte/src/routes/admin/admin-access.ts]: **Nit — comment slightly overstates where the user lands.**

"Returns the user to the login page" is only accurate for an unauthenticated visitor. For a logged-in non-admin who navigates to `/admin` in-app, `goto('/')` triggers the root layout's post-boot `$effect`, which will re-route them to `/game` (the session is still live). The same phrasing appears in `frontend.md` and the PR description.

The behaviour is correct and arguably more user-friendly than a hard logout, but the comment (and doc) should say something like "returns to root — the login page for unauthenticated users, or the game for an authenticated non-admin" to avoid confusing a future reader who expects a genuine logout.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: e2e-seed.sql [hunk 2]
```diff
@@ -68,3 +68,31 @@ ON CONFLICT ("ZoneId", "EnemyId") DO NOTHING;
...
+-- auto-attaches the seeded Admin role (id 1) to any account whose username marks it as an admin
+-- fixture ('e2eadmin…'). The admin-area tests can then sign in as a genuine admin while still using
+-- a unique account per test, keeping them parallel-safe.
+--
+-- e2e-only: this lives solely in this seed script, which never touches production/local databases.
+-- Idempotent: CREATE OR REPLACE + DROP ... IF EXISTS make re-running safe.
+CREATE OR REPLACE FUNCTION e2e_grant_admin_role() RETURNS TRIGGER AS $$
+BEGIN
+  IF NEW."Username" LIKE 'e2eadmin%' THEN
+    INSERT INTO "UserRoles" ("RolesId", "UsersId") VALUES (1, NEW."Id")
```
Comment (ginderjeremiah) [inline e2e-seed.sql]: **Nit — hardcoded role ID.**

The ID is currently correct (`ERole.Admin = 1` is explicitly set in the C# enum and seeded via `HasData`), and the comment calls it out. However, a lookup-by-name subquery would be more self-documenting and immune to a future enum renumber:

```sql
INSERT INTO "UserRoles" ("RolesId", "UsersId")
SELECT "Id", NEW."Id" FROM "Roles" WHERE "Name" = 'Admin'
ON CONFLICT DO NOTHING;
```

Not urgent — the value is stable — but worth considering if the seed file is touched again.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by ginderjeremiah — COMMENTED** [review]
## Review: frontend role-based access checks (#27)

**Intended verdict: Approve** — no blocking findings; two nits inline. (Submitting as `COMMENT` because GitHub won't accept `APPROVE` from the PR author.)

---

### Nits
- `UI/new-svelte/src/routes/admin/admin-access.ts:9` — JSDoc and `frontend.md` both say "returns to the login page", but `goto('/')` for an already-authenticated non-admin routes them back to `/game` via the boot gate's post-boot effect. The code is correct; only the comment/doc overstates it.
- `e2e-seed.sql:88` — role ID hardcoded as `1`; currently accurate (`ERole.Admin = 1` is stable), but a name-based subquery would be more self-documenting. See inline comment for the suggested form.

### Test coverage
Comprehensive. Every new module has a dedicated unit test file with good edge-case coverage:
- `token-store.test.ts` — `getRoles`: not-logged-in, non-JWT, missing claim, single-string, array, mixed-type array
- `roles.test.ts` — `hasRole`: no roles, matching role, non-matching role
- `screen-defs.test.ts` — `visibleScreens` for all four gate cases + `GAME_SCREENS` shape invariants
- `admin-access.test.ts` — grant, deny (toast + redirect), and role-specificity
- `NavSidebar.test.ts` — empty-group collapse when admin is filtered out
- e2e: `navigation.test.ts` adds a non-admin sidebar-hide assertion; `admin.test.ts` updated to use the new `createAdminAndStartGame` helper

The `ScreenDef` duplication between `NavSidebar.svelte` and `screen-defs.ts` is intentional and called out in the PR — the sidebar's local interface is a prop contract that doesn't need to know about role gating, and TypeScript structural typing makes it work cleanly.

### What's good
- Extracting `decodeAccessTokenPayload` removes the inline duplication that was in `getAccessTokenExpiry` — clean refactor.
- The `visibleScreens` pure function and `GAME_SCREENS` extraction make the gating rule independently testable without any component setup.
- Hydration-safe role snapshotting in `game/+page.svelte` (`onMount` snapshot, starting empty) mirrors the existing `booting`/`hydrated` boot-gate pattern and is well-commented.
- The e2e admin provisioning via a SQL trigger keyed on username prefix is elegant: creates genuine accounts through the real signup flow, parallel-safe (unique accounts per test), no brittle hand-seeded password hashes.
- The UX/security boundary is correctly documented: frontend gating is UX-only, `AdminRoleAuthorizationFilter` remains the real enforcement layer.

---
_Generated by [Claude Code](https://claude.ai/code)_
